### PR TITLE
Remove config dict for prefect ensemble

### LINKED ISSUE #1694: Integrate pydantic architecture to Ensemble (& PrefectEnsemble)
**Problem**
The evaluators `ert_shared/ensemble_evaluator/* and ert3/evaluator/*` were created in tandem, by different parties, with different strategies. and now, as a result, we have the [_build_ee_config ](https://github.com/equinor/ert/blob/master/ert3/evaluator/_evaluator.py#L108)function that builds an ensemble evaluator configuration dictionary from the ensemble and stages configurations ([_stages_config](https://github.com/equinor/ert/blob/master/ert3/config/_stages_config.py) and [_ensemble_config](https://github.com/equinor/ert/blob/master/ert3/config/_ensemble_config.py)).  This "middle man" function nor [PrefectEnsemble](https://github.com/equinor/ert/blob/8200a0da3b47606aad765107f4163d254340e285/ert3/evaluator/_evaluator.py#L281) (wich the output of  ee_config is given as arguments ) does not validate the values of the ensemble evaluator configuration. And even though PrefectEnsemble encapsulates its attributes, we cannot ensure the credibility of the input parameters, since it's a non-validated dictionary that can be mutated down the line prior to [PerfectEnsemble](https://github.com/equinor/ert/blob/master/ert_shared/ensemble_evaluator/prefect_ensemble.py#L101). This situation is redundant and non-optimal.

**Solution**
- Validation of the instance object fields should be introduced in the keychain and deligated to the object model itself. Pedantic is here your friend. 
- Instance object should be immutable, again pydantic is your friend
- middle man function `ee_config' should be removed.

**Code fundation**
[scratcher.zip](https://github.com/equinor/ert/files/6520707/scratcher.zip)

### LINKED ISSUE #1704: Create an Ensemble Evaluator API
In issue #1694 Ensemble and EnsemblePrefect models were replace for pydantic models, These models should not be referenced directly, but instead through an API. For more context read this [comment](https://github.com/equinor/ert/pull/1659#discussion_r632316528) An "API" non-pydantic models does exist and uses the builder pattern https://github.com/equinor/ert/blob/master/ert_shared/ensemble_evaluator/entity/ensemble.py#L706

Design and implement a `PrefectEnsemble` (& `Ensemble`) API that builds and evaluates a pydantic `Ensemble` model instance.

### PR DESCRIPTION
**Pull request by sondreso**
**Issue**
Resolves #1704 
Resolves #1694 (arguably)

**Approach**
Remove the prefect ensemble config and create ensemble using builders. Also set source at build time to ensure the ids are set correct.

### CODE REVIEW

**Comment by jondequinor** [comment]
🫀

File: ert3/evaluator/_evaluator.py [hunk 1]
```diff
@@ -110,85 +117,77 @@ def _build_ee_config(
...
+        .set_forward_model(
+            create_realization_builder().active(True).add_step(step_builder)
+        )
+    )
 
     inputs = _prepare_input(
         ensemble.storage_type, stage, parameters, storage_path, ensemble_size
     )
     outputs = _prepare_output(ensemble.storage_type, stage, storage_path, ensemble_size)
+    builder = builder.set_inputs(inputs).set_outputs(outputs)
```
Comment (sondreso) [inline ert3/evaluator/_evaluator.py:188]: The following is equivalent, but which do we prefer? :thinking: 

```suggestion
    builder.set_inputs(inputs).set_outputs(outputs)
```

Comment (jondequinor) [reply ert3/evaluator/_evaluator.py:188]: Latter? I think

Comment (sondreso) [reply ert3/evaluator/_evaluator.py:188]: I think I agree :+1:

File: tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+@pytest.mark.timeout(60)
+def test_run_prefect_ensemble(evaluator_config, poly_ensemble, ensemble_size):
+    evaluator = EnsembleEvaluator(poly_ensemble, evaluator_config, 0, ee_id="1")
+
+    with evaluator.run() as mon:
+        for event in mon.track():
+            if isinstance(event.data, dict) and event.data.get("status") in [
+                "Failed",
+                "Stopped",
```
Comment (jondequinor) [inline tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py:71]: While you're at it, drop the magic strings? ;)

File: tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            ]:
+                mon.signal_done()
+
+    assert evaluator._ensemble.get_status() == "Stopped"
```
Comment (jondequinor) [inline tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py:75]: ids.ENSEMBLE_…

File: tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    evaluator = EnsembleEvaluator(poly_ensemble, evaluator_config, 0, ee_id="1")
+
+    with evaluator.run() as mon:
+        cancel = True
+        for _ in mon.track():
+            if cancel:
+                mon.signal_cancel()
+                cancel = False
+
+    assert evaluator._ensemble.get_status() == "Cancelled"
```
Comment (jondequinor) [inline tests/ert_tests/ensemble_evaluator/prefect/test_prefect_ensemble.py:91]: magic string

**Comment by sondreso** [comment]
There is probably a bit more cleanup that could be done some places, but I think this is a natural place to stop. There is always more nitpicking one can do :slightly_smiling_face:

File: ert_shared/ensemble_evaluator/ensemble/prefect.py [hunk 5]
```diff
@@ -184,31 +153,28 @@ def get_flow(self, ee_id, real_range):
...
-                    outputs = self.config[ids.OUTPUTS][iens]
-                    max_retries = self.config.get(ids.MAX_RETRIES, DEFAULT_MAX_RETRIES)
+                    outputs = self._outputs[iens]
                     # Prefect does not allow retry_delay if max_retries is 0
-                    if max_retries == 0:
-                        retry_delay = None
-                    else:
-                        delay = self.config.get("retry_delay", DEFAULT_RETRY_DELAY)
-                        retry_delay = timedelta(seconds=delay)
+                    retry_delay = None if self._max_retries == 0 else self._retry_delay
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/ensemble/prefect.py:165]: I don't want this changed, but could this be done on `build()`?

File: ert_shared/ensemble_evaluator/ensemble/prefect.py [hunk 6]
```diff
@@ -282,27 +248,26 @@ def _evaluate(self, ee_config: EvaluatorServerConfig, ee_id):
...
-                self.config["outputs"][iens][output_name] = transmitter
+            i = i + self._real_per_batch
+        for iens, tasks in self._iens_to_task.items():
+            for task in tasks:
+                if isinstance(state_map[iens].result[task].result, Exception):
+                    raise state_map[iens].result[task].result
+                for output_name, transmitter in (
+                    state_map[iens].result[task].result.items()
+                ):
+                    self._outputs[iens][output_name] = transmitter
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/ensemble/prefect.py:270]: Such a win not having to mutate the config.

**Review by jondequinor — APPROVED** [review]
This is a huge improvement. :rocket:

File: ert_shared/ensemble_evaluator/ensemble/prefect.py (hunk 5 above)
Comment (sondreso) [reply ert_shared/ensemble_evaluator/ensemble/prefect.py:165]: Yes we could! Question is whether we want this prefect specific settings quirk to reside in the builder or be hidden inside the prefect ensemble entirely.

Comment (jondequinor) [reply ert_shared/ensemble_evaluator/ensemble/prefect.py:165]: If it's possible to make a decision on build time, then I think the responsibility should lie with the builder.

**Pull request merged**